### ISSUE TITLE
Move multiple elements at once using Render Position (move up/move down) in Sequence Elements Frame

### ISSUE THREAD

**Issue by mararunr** [smeighan/xLights#114]
So I've got a couple hundred elements and it would be nice to use the shift key to grab multiple elements (in a row) and move them up or down together (versus one at a time).  I tried both the CTRL and the SHIFT method trying to grab multiple elements at once but neither worked.

**Comment by AzGilrock** [comment]
Yeah I disabled multiple selection for now because the logic was getting crazy.  With multi-select you can select non-continguous items.  Trying to keep track of what was highlighted and re-highlighting the elements after the move was giving me a headache.  So at least you can move them for now.

**Comment by AzGilrock** [comment]
By the way I'd have to check with Dan but I'm not sure we are honoring this ordering for rendering yet.

**Comment by mararunr** [comment]
No worries at all.  Actually considered it maybe an enhancement myself.  Definitely something that can wait, just wanted to get it out there.  When I start new sequences now it's not an issue because I pull every element in at the beginning (and then use the eyeball icon to "hide").  It's just some of these test sequences from alpha testing that I'd like to hold on to, clean up, and finish.  Thanks for the reply Gil.

**Comment by AzGilrock** [comment]
Yeah I actually had it working with multi-select for Move Up and Delete Models.  But then Move Down didn't work...it was only moving the last one.

**Comment by AzGilrock** [comment]
I had a idea pop in my head while watching a movie and came back and coded this up in 15 minutes.  It's pretty cool cause you can select a group of items and leave gaps in-between and it still works.  It will quit allowing the move if the group hits the top of bottom of the list.

**Comment by mararunr** [comment]
Gil, sorry to be the bearer of bad news but there is another bug with this fix.  Now when you click on move up or move down, you lose visibility to the item being moved.  Let's say I have 200 models and I'm trying to move the one model that is currently in position 100 (move it up or down), I scroll down to that model, highlight it and then if I hit move up or move down, the frame with all the models listed immediately refreshes to the top of the list - I can't see model 100 anymore or it's movement.  On a side note, I hope you get this message as I can't figure out how to re-open a closed item.

**Comment by AzGilrock** [comment]
Well the items are still selected so you could just keep hitting move up.  What happens in the grid on the right?  I don't have enough elements so I have no vertical scrollbar on the right to see how it behaves.  I'd just write a new one cause this isn't a bug.  It's just a visual enhancement.  So if a user has a block of elements selected that don't fit in the window where should I place visibility after the move?  Some people might want to see the top and others will complain they want to see the bottom.

**Comment by mararunr** [comment]
I'll make a screencast - if I'm trying to move position 100 to position 37 I have to select position 100 and start hitting move up, then every once in a while I pause from hitting that button and have to scroll down in the model list to see how far up I've come, then if I touch the move up button, it refreshes to the top of the list again - very painful.

You are welcome to use my xlights*.xml files on my public drive to test with (over 200 models).

For a grouping, I would suggestion just making a single refresh scenario that works for either button (yes I understand some will not like no matter what is done).  My suggestion would be if you have say a group of 20 that you want to move, then , when you hit the move up or move down button, the model list frame displays the top of the list and the next 2 non-highlighted items above it so as you move up or down you can see both highlighted items and non-highlighted items so you know where you are in the list (like having GPS you know where you are and what's around you).

In my case if I just had a "sort by name" button, I'd be good because I name my models in such a way that if they sorted by name, they would be in proper order.

I'll try to screencast over lunch.....(and remember to slow it down before posting :-)

**Comment by AzGilrock** [comment]
Yeah part of the reason it's working the way it does is because the wxListCtrl object is a pain to swap items.  You would think they would have provided a method to do it but no it's up to every developer that uses that object to create their own way.  So to get it working quickly I used a method Dave had already written which deletes everything and completely rebuilds the list.  So that's why it refreshes back to the top.  If I can write a good item swap routine I could get it working more in line with what you're asking.  I tried some code I found on the net and there was some kind of bug when trying to read back the label and set it on the new item you gotta create to perform a swap.  It was ending up with a blank label.  I saw someone else complaining about the same thing so there could be a bug in that control so I might need to develop a workaround.

**Comment by mararunr** [comment]
Sorry Gil - and I hope you don't take my comments as complaining.  As I've said many times before you all have created a great tool even if you never touched another line of code - but even great can sometimes be made better.  Heck yesterday I started playing with papagayo just to try and test something new with xL4 :-)  (I built my first coro face smart pixel prop on Saturday.)

**Comment by dpitts** [comment]
I got tired of hardware for 10 minutes so I added some code to help keep the moving items in view. The up may need a little work but down seems good.

**Comment by mararunr** [comment]
Dave!  Welcome back from Falcon fun. :-)  Thank you both.  I'll skip the screencast and await for next release.

**Comment by AzGilrock** [comment]
Cool I took a look at the checkin and that looks good.   The only thing I was worried about is whether there was a performance issue by calling that PopulateModels call if a user has a boatload of models in the list.  That's why I thought if we created a method to swap the elements it might be faster than reconstructing the whole list if the list was really large but if mararunr has 200 elements that's probably a good enough test of performance.
